sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs

```diff
@@ -0,0 +1,226 @@
+                }
+                else
+                {
+                    // entityPath for a queue is "{QueueName}"
+                    processor = client.CreateSessionProcessor(entityPath, _options.ToSessionProcessorOptions());
+                }
+                processor.ProcessErrorAsync += _options.ExceptionReceivedHandler;
+                return processor;
+            });
+        }
+
+        internal ServiceBusClient CreateSessionClient(string connection) =>
+            GetOrCreateClient(connection);
+
+        private ServiceBusClient GetOrCreateClient(string connection)
+        {
+            var connectionInfo = ResolveConnectionInformation(connection);
+            if (connectionInfo.ConnectionString != null)
+            {
+                return _clientCache.GetOrAdd(
+                    connectionInfo.ConnectionString,
+                    (_) => new ServiceBusClient(connectionInfo.ConnectionString, _options.ToClientOptions()));
+            }
+            else
+            {
+                return _clientCache.GetOrAdd(
+                    connectionInfo.FullyQualifiedNamespace,
+                    (_) => new ServiceBusClient(connectionInfo.FullyQualifiedNamespace, connectionInfo.Credential, _options.ToClientOptions()));
+            }
+        }
+
+        private ServiceBusConnectionInformation ResolveConnectionInformation(string connection)
+        {
+            var connectionSetting = connection ?? Constants.DefaultConnectionStringName;
+            IConfigurationSection connectionSection = _configuration.GetWebJobsConnectionStringSection(connectionSetting);
+            if (!connectionSection.Exists())
+            {
+                // Not found
+                throw new InvalidOperationException($"Service Bus account connection string '{connectionSetting}' does not exist. " +
+                                                    $"Make sure that it is a defined App Setting.");
+            }
+
+            if (!string.IsNullOrWhiteSpace(connectionSection.Value))
+            {
+                return new ServiceBusConnectionInformation(connectionSection.Value);
+            }
+            else
+            {
+                string fullyQualifiedNamespace = connectionSection["fullyQualifiedNamespace"];
+                if (string.IsNullOrWhiteSpace(fullyQualifiedNamespace))
+                {
+                    // Not found
+                    throw new InvalidOperationException($"Connection should have an 'fullyQualifiedNamespace' property or be a " +
+                        $"string representing a connection string.");
+                }
+
+                TokenCredential credential = _componentFactory.CreateTokenCredential(connectionSection);
+                return new ServiceBusConnectionInformation(fullyQualifiedNamespace, credential);
+            }
+        }
+
+        private record ServiceBusConnectionInformation
```

Comment:
Since you're not doing equality comparisons nor creating new instances with partial updates,  I wonder if `struct` would be adequate, rather than `record`.

Reply:
Most of the time I do it for the ToString implementation :)